### PR TITLE
Fix inheritance in DataTree.copy()

### LINKED ISSUE #9454: Deep copying a datatree somehow duplicates data onto child nodes
### What happened?

Deep copying a DataTree object somehow duplicates data onto a child node (!)

### What did you expect to happen?

Copying to return an identical subtree.

### Minimal Complete Verifiable Example

```Python
In [15]: dt = DataTree(data=xr.Dataset(coords={'x': ('x', [0, 1])}))

In [16]: print(dt)
<xarray.DataTree>
Group: /
    Dimensions:  (x: 2)
    Coordinates:
      * x        (x) int64 16B 0 1
    Data variables:
        *empty*

In [17]: dt['c'] = DataTree()

In [18]: print(dt)
<xarray.DataTree>
Group: /
│   Dimensions:  (x: 2)
│   Coordinates:
│     * x        (x) int64 16B 0 1
│   Data variables:
│       *empty*
└── Group: /c

In [19]: dt.copy(deep=True)
Out[19]: 
<xarray.DataTree>
Group: /
│   Dimensions:  (x: 2)
│   Coordinates:
│     * x        (x) int64 16B 0 1
│   Data variables:
│       *empty*
└── Group: /c
        Dimensions:  (x: 2)
        Coordinates:
          * x        (x) int64 16B 0 1
        Data variables:
            *empty*
```

### Anything else we need to know?

_No response_

### Environment

`main`

### PR DESCRIPTION
**Pull request by shoyer**
Previously, we were copying parent coordinates/dimensions onto all child nodes. This is not obvious in the current repr, but you can see it from looking at the private `._node_coord_variables` and `._node_dims`.

To make the use of `_to_dataset_view()` little more obvious, I've added a required boolean `inherited` argument.

- [x] Closes #9454
- [x] Tests added

### CODE REVIEW

File: xarray/core/formatting.py [hunk 1]
```diff
@@ -1051,7 +1051,10 @@ def diff_datatree_repr(a: DataTree, b: DataTree, compat):
 def _single_node_repr(node: DataTree) -> str:
     """Information about this node, not including its relationships to other nodes."""
     if node.has_data or node.has_attrs:
-        ds_info = "\n" + repr(node._to_dataset_view(rebuild_dims=False))
+        # TODO: change this to inherited=False, in order to clarify what is
+        # inherited?
```
Comment (TomNicholas) [inline xarray/core/formatting.py:1055]: This is basically the same TODO that we mentioned in https://github.com/pydata/xarray/issues/8572#issuecomment-2218020742.

File: xarray/core/datatree.py [hunk 2]
```diff
@@ -490,30 +490,44 @@ def _dims(self) -> ChainMap[Hashable, int]:
...
-            #             "/": xr.Dataset({"a": (("x",), [1, 2])}),  # x has size 2
-            #             "/b/c": xr.Dataset({"d": (("x",), [3])}),  # x has size1
-            #         }
-            #     )
-            # However, they are fine for internal use cases, for align() or
-            # building a repr().
+        elif inherited:
+            # Note: rebuild_dims=False with inherited=True can create
+            # technically invalid Dataset objects because it still includes
+            # dimensions that are only defined on parent data variables, e.g.,
```
Comment (TomNicholas) [inline xarray/core/datatree.py:502]:
```suggestion
            # dimensions that are only defined on parent data variables (i.e. not present on any parent coordinate variables), e.g.,
```

File: xarray/core/datatree.py [hunk 3]
```diff
@@ -490,30 +490,44 @@ def _dims(self) -> ChainMap[Hashable, int]:
...
+            #     <xarray.DatasetView> Size: 0B
+            #     Dimensions:  (x: 2)
+            #     Dimensions without coordinates: x
+            #     Data variables:
+            #         *empty*
+            #
+            # The "x" dimension is still defined, even though there are no
+            # variables or coordinates. This is useful for use cases where we
+            # want to inherit everything from parents nodes, e.g., for align()
+            # and repr().
```
Comment (TomNicholas) [inline xarray/core/datatree.py:521]:
```suggestion
            # Notice the "x" dimension is still defined, even though there are no
            # variables or coordinates. 
            # Normally this is not supposed to be possible in xarray's data model, but here it is useful internally for use cases where we
            # want to inherit everything from parents nodes, e.g., for align()
            # and repr().
            # The user should never be able to see this dimension via public API.
```

**Review by TomNicholas — APPROVED** [review]
Thank you @shoyer ! That was quite a subtle bug.

My review comments are just about clarifying code comments.

File: xarray/core/formatting.py (hunk 1 above)
Comment (TomNicholas) [reply xarray/core/formatting.py:1055]: Raised https://github.com/pydata/xarray/issues/9464 to track this.

Comment (shoyer) [reply xarray/core/formatting.py:1055]: I made https://github.com/pydata/xarray/issues/9463 for this

Comment (TomNicholas) [reply xarray/core/formatting.py:1055]: Actuall #9463 is better!

File: xarray/core/datatree.py (hunk 3 above)
Comment (shoyer) [reply xarray/core/datatree.py]: applied

File: xarray/core/datatree.py (hunk 2 above)
Comment (shoyer) [reply xarray/core/datatree.py]: applied

**Comment by shoyer** [comment]
Thanks Tom for the comment suggestions!

**Pull request merged**